# Exercise 2.A Fundamentals of Pandas 

In [2]:
import numpy as np
import pandas as pd

## Series

In [3]:
# Creating Variables to hold plain Python Objects
aa = [1, 2, 3, 4]
ab = (5, 6, 7, 8)
ac = {9, 10, 11, 12}
ad = {
    "first": 13,
    "second": 14,
    "third": 15,
    "fourth": 16
}

print(f"List Object: {aa}\n")
print(f"Tuple Object: {ab}\n")
print(f"Set Object: {ac}\n")
print(f"Dictionary Object: {ad}\n")

List Object: [1, 2, 3, 4]

Tuple Object: (5, 6, 7, 8)

Set Object: {9, 10, 11, 12}

Dictionary Object: {'first': 13, 'second': 14, 'third': 15, 'fourth': 16}



In [4]:
# Creating Series from  Python Objects
# A Pandas Series is a one-dimensional labeled array
# pd.Series converts Python objects into Pandas Series
# Series are index-based
ser1 = pd.Series(aa)
print(f"List to Series: \n{ser1}\n")

ser2 = pd.Series(ab)
print(f"Tuple to Series: \n{ser2}\n")

# ser3 = pd.Series(ac)
# since Pandas can't convert a set to a series, it must be converted into a list or tuple
# I created a new variable called modifies_ser3 to convert the result of the list 
# conversion to a series
modified_ser3 = pd.Series(list(ac))
print(f"Set to Series: \n{modified_ser3}\n")

ser4 = pd.Series(ad)
print(f"Dictionary to Series: \n{ser4}\n")

List to Series: 
0    1
1    2
2    3
3    4
dtype: int64

Tuple to Series: 
0    5
1    6
2    7
3    8
dtype: int64

Set to Series: 
0     9
1    10
2    11
3    12
dtype: int64

Dictionary to Series: 
first     13
second    14
third     15
fourth    16
dtype: int64



In [5]:
ser5 = pd.Series(17, 18, 19, 20)

# The type error that arises is " Index(...) must be called with a collection of some kind"
# This means that you cannot covert a series without first assigning a collection of elements, 
# wether it be a list, dictionary, or tuple. The reason for this is that these collections
# have index positions; since Pandas series are index-based and literal values don't have
# indexes, Pandas can't convert them to a series.

TypeError: Index(...) must be called with a collection of some kind, 18 was passed

## DataFrames from Series

In [6]:
# Creating a DataFrame from Series
df1 = pd.DataFrame(ser1)
df1 # No print function is necessary, shows a cute table

,0
0,1
1,2
2,3
3,4


In [7]:
# Inserting New Columns to a DataFrame using .insert() method
# Format inserting from series and DataFrames is the same
# Format is index position of where you want to insert new column, column name, and column data
df1.insert(1, "ser2", ser2)
df1

,0,ser2
0,1,5
1,2,6
2,3,7
3,4,8


In [8]:
# .insert() only inserts one column at a time
# Inserting data from modified series 3
df1.insert(2, "ser3", modified_ser3)
df1

,0,ser2,ser3
0,1,5,9
1,2,6,10
2,3,7,11
3,4,8,12


In [9]:
df1.insert(3, "ser4", ser4)
df1
# ser4 is a dictionary-based Series, so its index is made up of string keys
# ('first', 'second', 'third', 'fourth') instead of numeric positions (0, 1, 2, 3).
# Since df1 uses a numeric index, pandas cannot align ser4's string index with it,
# resulting in NaN values across the entire column.

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN


In [10]:
# Using the loc attribute to add a new row
# In a DataFrame, you can add new rows by assigning to a loc index that doesn't exist yet
# by setting the data, it will just extend the dataframe. You can use either lists or tuples
df1.loc[4] = [5, 9, 12, 0]
df1

,0,ser2,ser3,ser4
0,1,5,9,NaN
1,2,6,10,NaN
2,3,7,11,NaN
3,4,8,12,NaN
4,5,9,12,0.0


## Reading Data into a DataFrame

In [11]:
# Using the read_csv() function to read contents of sakila database and convert to a DataFrame
df2 = pd.read_csv("sakila_example.csv")
df2

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE
...,...,...,...,...,...
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON


In [18]:
# Displaying the top 5 records
df2.head()

# Unless the data has a designated index column, the top row is always 0 because
# DataFrames are index-based, like Series

,title,description,category,language,actor_name
0,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,PENELOPE GUINESS
1,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,CHRISTIAN GABLE
2,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,LUCILLE TRACY
3,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,SANDRA PECK
4,ACADEMY DINOSAUR,A Epic Drama of a Feminist And a Mad Scientist...,Documentary,English,JOHNNY CAGE


In [21]:
# Displaying last 10 records
# Put 10 as an arugment, otherwise defualt is 5 rows
df2.tail(10)

,title,description,category,language,actor_name
252,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,MILLA KEITEL
253,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,GROUCHO DUNST
254,ATTACKS HATE,A Fast-Paced Panorama of a Technical Writer An...,Sci-Fi,English,BURT TEMPLE
255,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,UMA WOOD
256,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,RIP WINSLET
257,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,GARY PENN
258,ATTRACTION NEWTON,A Astounding Panorama of a Composer And a Fris...,New,English,CHRISTOPHER WEST
259,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,DUSTIN TAUTOU
260,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,ANGELA HUDSON
261,AUTUMN CROW,A Beautiful Tale of a Dentist And a Mad Cow wh...,Games,English,JAMES PITT


In [14]:
# Using .info() method to show summary information
df2.info()

<class 'pandas.DataFrame'>
RangeIndex: 262 entries, 0 to 261
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   title        262 non-null    str  
 1   description  262 non-null    str  
 2   category     262 non-null    str  
 3   language     262 non-null    str  
 4   actor_name   262 non-null    str  
dtypes: str(5)
memory usage: 10.4 KB


### Facts about **.info()** results

- The dataset contains 262 rows (entries) and 5 columns.
- The columns are: title, description, category, language, and actor_name.
- All columns contain 262 non-null values, meaning the dataset is clean and there are no missing values in the dataset.
- All columns are stored as string (str) data types, meaning the dataset mainly contains text data.
- The dataset uses approximately 10.4 KB of memory, which is relatively small.

In [22]:
# Using .describe() to understand distribution and stats of DataFrame
df2.describe()

,title,description,category,language,actor_name
count,262,262,262,262,262
unique,46,46,15,1,145
top,ARABIA DOGMA,A Touching Epistle of a Madman And a Mad Cow w...,Horror,English,OPRAH KILMER
freq,12,12,55,262,8


#### Useful Insights from `.describe()`
Since the sakila dataset contains text-based (object) columns, `.describe()` returns 
categorical statistics rather than numeric ones. It's showing count, unique values, the most 
frequent value (top), and its frequency (freq).

A useful insight is the top category and its frequency, which tells us which genre appears most often in the dataset. Another useful insight is the count and unique values, which show how many records exist and how much variety there is across titles and actors.

A less useful insight is the top actor_name, because it only shows which actor appears most frequently in the dataset — not how many unique movies they appear in or whether their films perform well commercially.

In [15]:
# Using the read_csv() function to read contents of northwind database and convert to a DataFrame
df3 = pd.read_csv("northwind_example.csv")
df3

,OrderID,ProductName,UnitPrice,Quantity
0,10248,Queso Cabrales,14.0,12
1,10248,Singaporean Hokkien Fried Mee,9.8,10
2,10248,Mozzarella di Giovanni,34.8,5
3,10249,Manjimup Dried Apples,42.4,40
4,10249,Tofu,18.6,9
...,...,...,...,...
495,10435,Mozzarella di Giovanni,27.8,10
496,10436,Gnocchi di nonna Alice,30.4,40
497,10436,Wimmers gute Semmelkndel,26.6,30
498,10436,Rhnbru Klosterbier,6.2,24


In [16]:
# Using .describe() to understand distribution and stats of DataFrame
df3.describe()

# 25%, 50%, and 75% are percentiles (quartiles) that show the distribution of the data
# 25% (first quartile) — 25% of the data falls below this value
# 50% (second quartile/median) — half the data falls below this value
# 75% (third quartile) — 75% of the data falls below this value
# Example: for Quantity, 50% of orders had a quantity of 20 or less, and 75% had 32 or less

,OrderID,UnitPrice,Quantity
count,500.000000,500.000000,500.000000
mean,10341.364000,23.371400,24.330000
std,54.442663,27.334436,18.069702
min,10248.000000,2.000000,1.000000
25%,10294.000000,10.400000,10.000000
50%,10341.000000,15.800000,20.000000
75%,10389.000000,27.800000,32.000000
max,10436.000000,210.800000,120.000000


#### Useful Insights from `.describe()`
Since the northwind dataset contains numerical columns, .describe() returns numerical statistics rather than categorical ones like the sakila database. It's showing the count, mean, standard deviation, minimum, maximum, and percentages across OrderID, UnitPrice, and Quantity.

A useful insight is the average quantity sold, minimum, and maximum quantity sold because it can provide indicators for top and lowest performing products.

A less useful insight is the OrderID statistics (mean, std, min, max) because OrderID is just an identifier, not a meaningful numeric value. Calculating statistics on OrderID doesn't reveal anything useful about the business data.

## Coding a DataFrame

In [24]:
# Creating a new variable and assiging it a dictionary
test_data = {
    'ID': [1, 2, 3, 4, 5, 6, 7, 8],
    'Name': ['Alicia', 'Bob', 'Charlie', 'Dominic', 'Eve', 'Frank', 'Grace', 'Heidi'],
    'Age': [25, np.nan, 22, 19, 31, 35, np.nan, 18],
    'City': ['NY', 'LA', 'Chicago', 'Houston', 'Phoenix', 'Boston', 'Austin', 'San Diego'],
    'Score': [85, 92, -1, 65, 78, np.nan, 55, 90]
}

In [25]:
# Creating a DataFrame from a dictionary
df4 = pd.DataFrame(test_data)
df4

,ID,Name,Age,City,Score
0,1,Alicia,25.0,NY,85.0
1,2,Bob,NaN,LA,92.0
2,3,Charlie,22.0,Chicago,-1.0
3,4,Dominic,19.0,Houston,65.0
4,5,Eve,31.0,Phoenix,78.0
5,6,Frank,35.0,Boston,NaN
6,7,Grace,NaN,Austin,55.0
7,8,Heidi,18.0,San Diego,90.0


In [26]:
# Exporting the DataFrame to a csv file
# You can export to csv, excel, or json file
# index=false is a boolean parameter that excludes the pandas index, making sure your 
# exported file doesn't have it as an extra unwanted column
df4.to_csv('test_data.csv', index=False)

The export was successful, I now have a test_data.csv file in my week-07 folder. It looks clean and the NaN values are just empty cells, no errors!